tests/unit-ext/test-ext-args.c

```diff
@@ -0,0 +1,152 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Unit test for jerry-ext/args.
+ */
+
+#include "jerryscript.h"
+#include "test-common.h"
+#include "jerryscript-ext/args.h"
+
+const char *test_source = (
+                           "var arg1 = true;"
+                           "var arg2 = 10.5;"
+                           "var arg3 = 'abc';"
+                           "var arg4 = function foo() {};"
+                           "test_validator(arg1, arg2, arg3, arg4);"
+                           "arg1 = new Boolean(true);"
+                           "arg3 = new String('abc');"
+                           "test_validator(arg1, arg2, arg3);"
+                           "arg2 = new Number(10.5);"
+                           "test_validator(arg1, arg2, arg3);"
+                           "test_validator(arg1, 10.5, 'abcdef');"
+                           );
+
+static int validator_count = 0;
+
+/**
+ * The handler should have following arguments:
+ *   this: Ignore.
+ *   arg1: Bool.
+ *   arg2: Number. It must be strict primitive number.
+ *   arg3: String.
+ *   arg4: function. It is an optional argument.
+ *
+ */
+static jerry_value_t
+test_validator_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
+                        const jerry_value_t this_val, /**< this value */
+                        const jerry_value_t args_p[], /**< arguments list */
+                        const jerry_length_t args_cnt) /**< arguments length */
+{
+
+  bool arg1;
+  double arg2 = 0.0;
+  char arg3[5];
+  jerry_value_t arg4 = jerry_create_undefined ();
+
+  jerryx_arg_t mapping[] =
+  {
+    /* ignore this */
+    jerryx_arg_ignore (),
+    /* 1st argument should be boolean */
+    jerryx_arg_boolean (&arg1, JERRY_ARG_CONVERSION_YES, JERRY_ARG_REQUIRED),
+    /* 2nd argument should be strict number */
+    jerryx_arg_number (&arg2, JERRY_ARG_CONVERSION_NO, JERRY_ARG_REQUIRED),
+    /* 3th argument should be string */
+    jerryx_arg_string (arg3, 5, JERRY_ARG_CONVERSION_YES, JERRY_ARG_REQUIRED),
+    /* 4th argument should be function, and it is optional */
+    jerryx_arg_function (&arg4, JERRY_ARG_OPTIONAL)
+  };
+
+  jerry_value_t is_ok = jerryx_validate_and_assign_all (this_val,
+                                                        args_p,
+                                                        args_cnt,
+                                                        mapping,
+                                                        5);
```

Comment:
@jiangzidong I'd like to know whether we succeeded in creating an API that actually also saves space compared to doing manual validation.

I suspect that for very simple manual validations (i.e. only needing `jerry_value_is_number(arg[0])` or something like that), it adds more code space when using `jerryx_validate_and_assign_all`.

Could you analyze when (i.e. how many manual validation function calls) you "break even" with this utility (on ARM Thumb2 and also x86)?

I also wonder if it's better (from a code size perspective at the call-site), to return a `bool` from the functions (and add an `jerry_value_t *out_error`). I think you have to call `jerry_value_has_error_flag` every time your call `jerryx_validate_and_assign_all` now. `jerry_value_has_error_flag` is an actual function call, so it's a bit expensive.